### ISSUE TITLE
Hypothesis tests are not deterministic

### ISSUE THREAD

**Issue by ezyang** [pytorch/pytorch#25248]
I suspect that Hypothesis tests in the form

```
    @given(X=hu.tensor(shapes=hu.array_shapes(min_dims=3, max_dims=4,
                                              min_side=1, max_side=10),
                       qparams=hu.qparams()),
           output_size_h=st.integers(1, 10),
           output_size_w=st.integers(1, 10))
    def test_adaptive_avg_pool2d(self, X, output_size_h, output_size_w):
```

are not deterministic. This is because #25097 is flaky; it should deterministically fail if we are seeding properly. We need to make sure that we actually run our tests deterministically; and furthermore, they must be stable with respect to other changes in the file.

cc @jerryzh168 @jianyuh @dzhulgakov @raghuramank100 @jamesr66a @vkuzo

**Comment by suo** [comment]
This flakiness is causing noise on my (and I assume everyone else's) PRs. @zafartahirov, @jerryzh168, can we disable the flaky tests ASAP while you investigate? We should not inflict known flakiness on everyone.

**Comment by z-a-f** [comment]
> @jerryzh168 There is still utility to hypothesis even if the test is derandomized. Think of it as a way to generate a large number of (deterministic) test cases under some parameter, without having to write them all manually. That's useful!
> 
> Randomized tests in CI is a recipe flaky tests, and flaky tests are unacceptable. We have plenty of unintended flakiness, let's not add intentional flakiness too.

The problem is that making the hypothesis always run the same set of tensors defies its purpose. I wanted to make sure that the hypothesis performs guided random testing, and finds failing corner cases over time. An alternative would be testing exhaustively, which will be extremely slow.

Ideas?

**Comment by suo** [comment]
We can rotate the seed manually once in a while. This will increase coverage while eliminating non-determinism. It is very important that our tests be deterministic, as flakiness affects all PRs and parts of the project.

**Comment by jamesr66a** [comment]
@z-a-f I don't think it's appropriate to do randomized testing on the main-line PR CI system. That would make sense if we're doing some sort of quantization-specific contbuild where the signals go directly to us, but it is a hard sell to say that causing failures on random unrelated PRs is acceptable.

Proposal:
1) Fully fix hypothesis seed in OSS CI
2) Spin up internal testing wil randomized tests to catch errors over time

**Comment by jerryzh168** [comment]
Yeah hypothesis tests are derandomized on OSS CI right now.
For 2). I think we talked about this before, is xla doing similar things right now? @ailzhang

**Comment by jerryzh168** [comment]
Talked with @ailzhang offline, she mentioned we can use `slowTest`, @ezyang can you share something about that? when does the tests marked with `slowTest` run?

**Comment by z-a-f** [comment]
> 1. Fully fix hypothesis seed in OSS CI
> 2. Spin up internal testing wil randomized tests to catch errors over time

I agree that CI should only run deterministic tests. Maybe we can get extra machines to have third option: run randomized tests (or any monte carlo methods) continuously?

**Comment by ezyang** [comment]
@jerryzh168 slow tests run in the "slow-test" job. This job runs by default on all PRs.

If your randomized tests aren't catching errors at a single seed, you need to run more instances of your randomized tests. Also, when you discover regressions, you should hard code those examples as test cases that always get run.

**Comment by jerryzh168** [comment]
right now we are abusing hypothesis to do unittest actually, and we are wondering how hard is it to have a separate CI service, which does not block the PRs to land, that can run randomized hypothesis tests?

**Comment by ezyang** [comment]
This is what nightlies are for. You can ask @kostmo or @mingbowan for more information about how these are setup.

**Comment by raghuramank100** [comment]
How about having a mix of hypothesis tests: Deterministic, but focus on testing corner cases and randomized tests, with a fixed seed: Deterministic, but test more typical cases?

________________________________

**Comment by jerryzh168** [comment]
We can have two types of tests:
1). unit tests that test typical cases and run on every PR.
2). randomized hypothesis tests that test corner cases and runs separately

**Comment by vkuzo** [comment]
adding to triage review since there have been a lot of complaints of hypothesis being a net negative for quantization op tests

**Comment by vkuzo** [comment]
after team discussion, we plan to stop using hypothesis in quantization unit tests

**Comment by Zac-HD** [comment]
Hi all 👋 - I'm a Hypothesis core dev, and generally happy to help out on OSS projects.  Two things that spring to mind:

1. @rsokl noticed that you're not [using `hypothesis.register_random()`](https://hypothesis.readthedocs.io/en/latest/details.html#making-random-code-deterministic) - if your tests are influenced by a PRNG that Hypothesis hasn't been shown how to manage, that could be the source of flakiness or nondeterminism.

2. I've been working on https://hypofuzz.com/, to get the most out of a "throw more compute at the problem" approach to nightly or ongoing random testing.  Let me know if you're interested in trying it out; it sounds like you're at least experimenting with the workflow it's designed for.

**Comment by vkuzo** [comment]
@Zac-HD , would you have any tips on how one can configure Hypothesis to generate tensors with values similar to something like `torch.randn`? For testing quantization kernels, we often want input tensors which look comparable to what one would expect inside a neural network.  When generating inputs with Hypothesis we usually get a lot of inputs with extreme values, without ever seeing realistic inputs.  This enables us to test for edge cases, but not for the common case.

For example, let's say we'd like to generate 4x1 matrices to feed into a kernel.  Currently we may be getting matrices like [0, 0, 0, 0], [0, 0, 0, 1], [0, 0, 0, max_float], ...  from hypothesis.  We're hoping to figure out how to also get matrices like [0.05, -0.24, 0.13, 1.31] in the generated input set, without losing the edge casey ones.

**Comment by Zac-HD** [comment]
Good question!

The first thing I'd recommend is using the `@example()` decorator to feed in some hand-written (or randomly-generated-with-fixed-seed) normal-looking examples.  That's a quick and clean patch, and as a general rule there are very few bugs in the part of your input space which is 'looks normal' on all dimensions.

Unfortunately for this specific case, Hypothesis doesn't and basically can't support specific distributions of values - there are just too many heuristics and feedbacks to do it, and the interactions with shrinking are pretty complicated too.  *However*, you can generally focus it on particular areas by customizing the strategy - in this case by taking the union of strategies for edge-casey and a normal-looking matrices.  e.g.:

```python
matrices = st.one_of([
    tensors(shape, dtype, elements=elems)
    for elems in (st.floats(), st.floats(-10, 10), st.floats(-2, 2))  # or combinations of such
])
```

This is pretty ugly, but it mostly works.

The other major approach is to use Hypothesis to chose shapes and dtypes, but `torch.random` (or whatever) to select the elements.  On the upside that gives you normal-looking elements and maybe a perf boost; on the downside it doesn't give you the central-limit-theorem-defying edge cases or shrinking or feedback-guided testing.  In principle we could "assimilate" externally-generated elements for later shrinking, but while most of the components are [on our issue tracker](https://github.com/HypothesisWorks/hypothesis/issues/3086) it would be a 2-4 month project and nobody with the time and skills has yet shown up to take it on.

(I also have a patch in the works which will give arrays of up to ~10 elements more distinct elements, but we'll see whether or how much that helps here)

**Comment by z-a-f** [comment]
Closing this -- the consensus is to remove hypothesis from the quantization tests. Please, reopen if still needed
